File: test/integ/TypeAnalysisTransformTest.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.redextest;

class Base {
  int getVal() { return 0; }
}

class SubOne extends Base {
  @Override
  int getVal() {
    return 1;
  }
}

class SubTwo extends Base {
  @Override
  int getVal() {
    return 2;
  }
}

/*
 * Anonymous class call back example
 */
interface Callback {
  void hook();
}

class RenderView {
  Callback mCallback = new Callback() {
    @Override
    public void hook() {
      checkBaseField();
    }
  };
  Base mBase;

  RenderView() { mBase = new Base(); }

  int checkBaseField() {
    if (mBase != null) {
      return 2;
    }
    return 1;
  }
}

/*
 * Primitive arrays
 */
class ReactNode {
  final boolean[] mIsCool = new boolean[4];

  void setCool(final int i, final boolean isCool) { mIsCool[i] = isCool; }

  String getCool(final int i) {
    if (mIsCool[i]) {
      return "cool";
    }
    return "not cool";
  }
}

class TestRemoveRedundantNullChecks {

  Base mField1;

  TestRemoveRedundantNullChecks() {
    int v = checkEQZInitReachable(mField1);
    v += checkEQZInitReachableGetField();
    mField1 = new Base();
  }

  public Base getSubOne() { return new SubOne(); }

  public Base getSubTwo() { return new SubTwo(); }

  public int checkEQZNullArg(Base b) {
    int i = 0;
    if (b != null) {
      i += b.getVal();
    }
    return i;
  }

  public int checkEQZNotNullArg(Base b) {
    int i = 0;
    if (b != null) {
      i += b.getVal();
    }
    return i;
  }

  public int checkNEZNullArg(Base b) {
    int i = 0;
    if (b == null) {
      i += b.getVal();
    }
    return i;
  }

  public int checkNEZNotNullArg(Base b) {
    int i = 0;
    if (b == null) {
      i += b.getVal();
    }
    return i;
  }

  private int checkEQZInitReachable(Base b) {
    int i = 0;
    if (b != null) {
      i += b.getVal();
    }
    return i;
  }

  private int checkEQZInitReachableGetField() {
    int i = 0;
    if (this.mField1 != null) {
      i += this.mField1.getVal();
    }
    return i;
  }

  static void main() {
    TestRemoveRedundantNullChecks t = new TestRemoveRedundantNullChecks();
    Base b = null;
    t.checkEQZNullArg(b);
    b = new Base();
    t.checkEQZNotNullArg(b);

    b = null;
    t.checkNEZNullArg(b);
    b = new Base();
    t.checkNEZNotNullArg(b);

    RenderView rv = new RenderView();
    rv.checkBaseField();

    ReactNode node = new ReactNode();
    node.setCool(0, true);
    node.setCool(1, false);
    String isCool = node.getCool(1);
  }
}

class TestRemoveRedundantTypeChecks {
  Base mField1;

  TestRemoveRedundantTypeChecks() {
    mField1 = new SubOne();
  }

  public int checkInstanceOfBaseNullArg(Base b) {
    int i = 42;
    if (b instanceof Base) {
      i += b.getVal();
    }
    return i;
  }

  public int checkInstanceOfBaseNotNullArg(Base b) {
    int i = 42;
    if (b instanceof Base) {
      i += b.getVal();
    }
    return i;
  }

  public int checkInstanceOfSubOneArg(Base b) {
    int i = 42;
    if (b instanceof SubOne) {
      i += ((SubOne) b).getVal();
    }
    return i;
  }

  public int checkInstanceOfSubTwoArg(Base b) {
    int i = 42;
    if (b instanceof SubTwo) {
      i += ((SubTwo) b).getVal();
    }
    return i;
  }

  public int checkInstanceOfNullableField() {
    int i = 42;
    Base b = mField1;
    if (b instanceof SubTwo) {
      i += ((SubTwo) b).getVal();
    }
    return i;
  }

  static void main() {
    TestRemoveRedundantTypeChecks t = new TestRemoveRedundantTypeChecks();
    Base b = null;
    t.checkInstanceOfBaseNullArg(b);
    b = new Base();
    t.checkInstanceOfBaseNotNullArg(b);

    b = new SubOne();
    t.checkInstanceOfSubOneArg(b);
    t.checkInstanceOfSubTwoArg(b);

    // null assignment making the field nullable
    t.mField1 = null;
    t.checkInstanceOfNullableField();
  }
}

File: test/integ/TypeAnalysisTransformTest.cpp
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include <gtest/gtest.h>

#include "DexClass.h"
#include "GlobalTypeAnalysisPass.h"
#include "TypeAnalysisTestBase.h"

using namespace type_analyzer;
using namespace type_analyzer::global;

using TypeSet = sparta::PatriciaTreeSet<const DexType*>;

class TypeAnalysisTransformTest : public TypeAnalysisTestBase {};

TEST_F(TypeAnalysisTransformTest, RemoveRedundantNullCheckTest) {
  auto scope = build_class_scope(stores);
  set_root_method(
      "Lcom/facebook/redextest/TestRemoveRedundantNullChecks;.main:()V");

  auto gta = new GlobalTypeAnalysisPass();
  gta->get_config().transform.remove_redundant_null_checks = true;
  std::vector<Pass*> passes{gta};
  run_passes(passes);

  {
    auto meth_check_null_arg =
        get_method("TestRemoveRedundantNullChecks;.checkEQZNullArg",
                   "Lcom/facebook/redextest/Base;", "I");
    auto code = meth_check_null_arg->get_code();
    const auto& insns = InstructionIterable(code);
    for (auto& mie : insns) {
      EXPECT_NE(mie.insn->opcode(), OPCODE_IF_EQZ);
    }
  }
  {
    auto meth_check_not_null_arg =
        get_method("TestRemoveRedundantNullChecks;.checkEQZNotNullArg",
                   "Lcom/facebook/redextest/Base;", "I");
    auto code = meth_check_not_null_arg->get_code();
    const auto& insns = InstructionIterable(code);
    for (auto& mie : insns) {
      EXPECT_NE(mie.insn->opcode(), OPCODE_IF_EQZ);
    }
  }
  {
    auto meth_check_null_arg =
        get_method("TestRemoveRedundantNullChecks;.checkNEZNullArg",
                   "Lcom/facebook/redextest/Base;", "I");
    auto code = meth_check_null_arg->get_code();
    const auto& insns = InstructionIterable(code);
    for (auto& mie : insns) {
      EXPECT_NE(mie.insn->opcode(), OPCODE_IF_NEZ);
    }
  }
  {
    auto meth_check_null_arg =
        get_method("TestRemoveRedundantNullChecks;.checkNEZNotNullArg",
                   "Lcom/facebook/redextest/Base;", "I");
    auto code = meth_check_null_arg->get_code();
    const auto& insns = InstructionIterable(code);
    for (auto& mie : insns) {
      EXPECT_NE(mie.insn->opcode(), OPCODE_IF_NEZ);
    }
  }
  {
    auto meth_check_null_arg =
        get_method("TestRemoveRedundantNullChecks;.checkEQZInitReachable",
                   "Lcom/facebook/redextest/Base;", "I");
    auto code = meth_check_null_arg->get_code();
    const auto& insns = InstructionIterable(code);
    bool found_if_eqz = false;
    for (auto& mie : insns) {
      if (mie.insn->opcode() == OPCODE_IF_EQZ) {
        found_if_eqz = true;
      }
    }
    EXPECT_TRUE(found_if_eqz);
  }
  {
    auto meth_check_null_arg = get_method(
        "TestRemoveRedundantNullChecks;.checkEQZInitReachableGetField", "",
        "I");
    auto code = meth_check_null_arg->get_code();
    const auto& insns = InstructionIterable(code);
    bool found_if_eqz = false;
    for (auto& mie : insns) {
      if (mie.insn->opcode() == OPCODE_IF_EQZ) {
        found_if_eqz = true;
      }
    }
    EXPECT_TRUE(found_if_eqz);
  }
  {
    auto meth_check_base_field =
        get_method("RenderView;.checkBaseField", "", "I");
    auto code = meth_check_base_field->get_code();
    const auto& insns = InstructionIterable(code);
    bool found_if_eqz = false;
    for (auto& mie : insns) {
      if (mie.insn->opcode() == OPCODE_IF_EQZ) {
        found_if_eqz = true;
      }
    }
    EXPECT_TRUE(found_if_eqz);
  }
  {
    auto meth_check_primitive_array_field =
        get_method("ReactNode;.getCool", "I", "Ljava/lang/String;");
    auto code = meth_check_primitive_array_field->get_code();
    const auto& insns = InstructionIterable(code);
    bool found_if_eqz = false;
    for (auto& mie : insns) {
      if (mie.insn->opcode() == OPCODE_IF_EQZ) {
        found_if_eqz = true;
      }
    }
    EXPECT_TRUE(found_if_eqz);
  }
}

TEST_F(TypeAnalysisTransformTest, TestRemoveRedundantTypeChecks) {
  auto scope = build_class_scope(stores);
  set_root_method(
      "Lcom/facebook/redextest/TestRemoveRedundantTypeChecks;.main:()V");

  auto gta = new GlobalTypeAnalysisPass();
  gta->get_config().transform.remove_redundant_type_checks = true;
  std::vector<Pass*> passes{gta};
  run_passes(passes);

  {
    auto meth_check_null_arg =
        get_method("TestRemoveRedundantTypeChecks;.checkInstanceOfBaseNullArg",
                   "Lcom/facebook/redextest/Base;", "I");
    auto code = meth_check_null_arg->get_code();
    const auto& insns = InstructionIterable(code);
    bool found_const_0 = false;
    for (auto& mie : insns) {
      auto insn = mie.insn;
      EXPECT_NE(insn->opcode(), OPCODE_INSTANCE_OF);
      if (insn->opcode() == OPCODE_CONST && insn->get_literal() == 0) {
        found_const_0 = true;
      }
    }
    EXPECT_TRUE(found_const_0);
  }
  {
    auto meth_check_not_null_arg = get_method(
        "TestRemoveRedundantTypeChecks;.checkInstanceOfBaseNotNullArg",
        "Lcom/facebook/redextest/Base;", "I");
    auto code = meth_check_not_null_arg->get_code();
    const auto& insns = InstructionIterable(code);
    bool found_const_1 = false;
    for (auto& mie : insns) {
      auto insn = mie.insn;
      EXPECT_NE(insn->opcode(), OPCODE_INSTANCE_OF);
      if (insn->opcode() == OPCODE_CONST && insn->get_literal() == 1) {
        found_const_1 = true;
      }
    }
    EXPECT_TRUE(found_const_1);
  }
  {
    auto meth_check_sub_one =
        get_method("TestRemoveRedundantTypeChecks;.checkInstanceOfSubOneArg",
                   "Lcom/facebook/redextest/Base;", "I");
    auto code = meth_check_sub_one->get_code();
    const auto& insns = InstructionIterable(code);
    bool found_const_1 = false;
    for (auto& mie : insns) {
      auto insn = mie.insn;
      EXPECT_NE(insn->opcode(), OPCODE_INSTANCE_OF);
      if (insn->opcode() == OPCODE_CONST && insn->get_literal() == 1) {
        found_const_1 = true;
      }
    }
    EXPECT_TRUE(found_const_1);
  }
  {
    auto meth_check_sub_two =
        get_method("TestRemoveRedundantTypeChecks;.checkInstanceOfSubTwoArg",
                   "Lcom/facebook/redextest/Base;", "I");
    auto code = meth_check_sub_two->get_code();
    const auto& insns = InstructionIterable(code);
    bool found_const_0 = false;
    for (auto& mie : insns) {
      auto insn = mie.insn;
      EXPECT_NE(insn->opcode(), OPCODE_INSTANCE_OF);
      if (insn->opcode() == OPCODE_CONST && insn->get_literal() == 0) {
        found_const_0 = true;
      }
    }
    EXPECT_TRUE(found_const_0);
  }
  {
    auto meth_check_nullable_field = get_method(
        "TestRemoveRedundantTypeChecks;.checkInstanceOfNullableField", "", "I");
    auto code = meth_check_nullable_field->get_code();
    const auto& insns = InstructionIterable(code);
    bool found_instance_of = false;
    for (auto& mie : insns) {
      auto insn = mie.insn;
      if (insn->opcode() == OPCODE_INSTANCE_OF) {
        found_instance_of = true;
      }
    }
    EXPECT_TRUE(found_instance_of);
  }
}
